XCloud Fix

ChuckL opened the pull request:
Finished up fixing the server.

With the blog post to get the old xcloud working I added a button to use the old xcloud at the xcloud home page.

I would still recommend upgrading to the new xcloud since it's using https for everything.
Also I change the preferences so xcloud doesn't upload its own settings since it was uploading the user's password in md5 hash insecurely.

The file is also modular so it no longer requires the xkit god object's page get function.

XCloud source is https://github.com/ChuckL/XCloud
XCloud-Portal https://github.com/ChuckL/XCloud-Portal | https://portal.new-xkit.com/

XCloud portal is for user's xcloud management.  It will allow the user to check if their extensions are there before upgrading to new xkit.

ChuckL commented:
Oh and this fixes #53

bvtsang commented:
Don't forget to bump up the version.

bvtsang commented on Extensions/xcloud.js:
```diff
@@ -12,6 +12,15 @@ XKit.extensions.xcloud = new Object({
 
+
+	get_xcloud_url: function() {
+		if (XKit.extensions.xcloud.useoldserver === true) {
```
`if (XKit.extensions.xcloud.useoldserver)`

There's other places in this code that can use this convention as well.

ChuckL replied:
Should we use that convention?
I'm not sure if I want true-ish values to be true.

bvtsang replied:
Are the values expected to be not boolean? If that's the case, then ignore me.

bvtsang commented on Extensions/xcloud.js:
```diff
@@ -190,42 +208,64 @@ XKit.extensions.xcloud = new Object({
-					XKit.window.show("Can't connect to server","XKit was unable to contact XCloud servers.<br/>Please try again or <a href=\"http://xkit-extension.tumblr.com/ask\">send a bug report</a>.","error","<div id=\"xkit-close-message\" class=\"xkit-button default\">OK</div>");
-					return;
-				}
+					if (mdata.server_down === true) {
```
`if (mdata.server_down)`

bvtsang commented on Extensions/xcloud.js:
```diff
@@ -403,51 +466,67 @@ XKit.extensions.xcloud = new Object({
+				if (data.errors === "false") {
 
-			} else {
+					// GM_deleteAllValues(function() {
```
Remove unneeded comment.

bvtsang commented on Extensions/xcloud.js:
```diff
@@ -594,28 +673,28 @@ XKit.extensions.xcloud = new Object({
+				bufView[i] = str.charCodeAt(i);
 		}
-		return ab;
+		return buf;*/
```
Remove unneeded comments.

ChuckL commented:
Found a bug where it was not saving to XKit storage.
Also fixed an issue where after restore it would not keep the user signed in.

ThePsionic commented:
364 commits... what did you even do

0xazure commented:
The git history in this one is going to need some cleanup before merging.  All of the commits before 96e07c319f9f5f100b60574c1617d893370f0c41 are from `gh-pages` and should **not** be merged.

What branch was this work branched from, originally?

ChuckL commented:
I branched from master, however after I ran the extension script I goofed and did some work on gh-pages and merged back.  I'll work on fixing that up.

ChuckL commented:
Commit history should be fixed.

nightpool commented:
What's our deployment plan here? Is this ready to go/do we trust the new servers not to fall apart under load/should we stagger deployment somehow? 

@ChuckL it would be nice if you could squash some of these commits into logical changesets. thx!

hobinjk commented on Extensions/xcloud.js:
```diff
@@ -813,23 +902,23 @@ XKit.extensions.xcloud = new Object({
-		var xl;
-
-		var rotateLeft = function (lValue, iShiftBits) {
+		  // http://kevin.vanzonneveld.net
```
The section starting here has a large amount of introduced whitespace errors (two spaces instead of \t)

hobinjk commented on Extensions/xcloud.js:
```diff
@@ -12,6 +12,15 @@ XKit.extensions.xcloud = new Object({
 
+
+	get_xcloud_url: function() {
+		if (XKit.extensions.xcloud.useoldserver) {
```
`useoldserver` should be given a default value above.

ChuckL commented:
@nightpool   Done!

I currently have XCloud hosted at AWS on a EC2 t2.micro instance at https://cloud.new-xkit.com

Unfortunately I have no idea how much the load will be, however the CPU should be able to bust to suite it for a little while until we get a baseline.

From there if it's too much for a t2.micro we can upgrade to what will suite it.
My funds will probably only allow it to go to a t2.medium.  If we need to go past a t2.micro I'll probably look into using AWS Lambda instead.  But hey, I'm hoping I wrote the server slim enough to handle the workload on a micro instance.

As for datastorage, I'm not worried about it.  It's storing user creds in DynamoDB and payloads in S3 which are both pretty cheap for what we're doing.

ChuckL commented:
I suppose it would be good to get a heads up on gitter before it's merged in so I can baby sit it for a bit.

ChuckL commented:
Also standardized url endpoints.  This is just in case the server can't handle it and we have to switch to Lambda.

ChuckL commented:
I'

ChuckL commented:
I'll just work on switching to Lambda now to alleviate server concerns.

nightpool commented:
let's not let perfect be the enemy of great. I'm more concerned about storage then capacity, but if we want to try some sort of limited release I'd be down.

ChuckL commented:
Fair enough.  I think migrating to Lambda won't be too much work.  But if we want to release before I am able to then I would be perfectly fine with that.

For storage I'm not concerned.  DynamoDB and S3 should just give us everything that we ask for.
https://aws.amazon.com/dynamodb/pricing/
https://aws.amazon.com/s3/pricing/

Dynamo is setup with 20 read/write units.

ChuckL commented:
Lamba with bcrypt is being as slow as molasses on a cold Canadian day so at the moment that's not feasible.  I'll probably keep trying it however just to get XCloud up and running we can just use the current EC2 instance it's deployed to.

ChuckL commented:
Figured out what was wrong with Lambda after a week.  Now pointing to AWS Lambda instead of EC2.
Lambda will also give us built in metrics.

hobinjk commented:
This LGTM as soon as there is a post announcing the changes and detailing the steps required to switch back to the old server.

Wolvan commented:
@hobinjk I think the old servers are already an automatic fallback as far as I understand @ChuckL in the gitterchats

ChuckL commented:
The automatic fallback is for when user's are downloading extensions from xcloud.

If the user downloads their payload and if has an extension that's not on their xkit, it will first try to install the extension from our github server and then if it can't find it, fallback and download from Atesh's server.

but @hobinjk about the switch back to the old server.  i'm going to make a quick change to ensure users updating will still hit the old server then if you're logging in you can choose between them.

I also have a migration guide on the cloud portal I was working on. https://portal.new-xkit.com/assets/migration/xcloud-migration.html

Should I move all of that stuff to the new-xkit org?

ChuckL commented:
@hobinjk 
https://portal.new-xkit.com/assets/documents/xcloud-annoucement/

Draft of announcement post.

hobinjk commented:
@ChuckL There's a minor misspelling in the URL: "annoucement". The post itself looks great, although a picture of where "Use old XCloud Servers" is located would be nice to have.

I think it should now be possible for you to transfer XCloud to new-xkit. If it doesn't work then I can temporarily promote you to be an Owner.

ChuckL commented:
Updated and added a picture.
https://portal.new-xkit.com/assets/documents/xcloud-announcement/index.html

XCloud is now on new-xkit.
XCloud portal is the gh-pages for the repo.

I was thinking the page linked is just a sample of the blog post.  Someone would need to transcribe that to Tumblr.

bvtsang commented:
Nitpicks:

- Remove the period in "What this means to you." because it's inconsistent with the other headings
- For users on new XKit _who_ couldn't get XCloud to work
- For users on new XKit _who were_ using XCloud perfectly fine
- Remove the period after "page" in 'Just make sure you check "Use the old XCloud Servers" on the welcome page. to use the old XCloud servers.'

ChuckL commented:
Polished document with btsang's comments.

gloomykid commented:
I've tried using the new XCloud today and I just noticed that the header options are always on "reset" mode when synced and restored. For example: I am not using the blog bubbles of all my sideblogs, and my header is fixed, but when I restored after a sync, it's neither fixed and my sideblogs have icons bubbles on the header.

ChuckL commented:
Humm, odd.
Could you go to https://portal.new-xkit.com -> login -> hit status.
Then check if there are any invalid payloads?

ChuckL commented:
Moving this to in progress since there are cases of data being lost.

gloomykid commented:
I just tried the status thing and the only invalid in my entire list was the header options.

![ss 2015-09-04 at 06 44 42](https://cloud.githubusercontent.com/assets/13604430/9689267/3ef49f22-5335-11e5-8b01-78ecf56d7116.png)

The rest of the list is okay.

ChuckL commented:
Awesome thanks!  That verifies my suspicions.  

I found a bug where if there's unicode in the extension it has a chance of corrupting the payload due to how base64 encodes and decodes.  This bug seems to also be a problem in how the old xcloud operated too.

I'll be working on a fix shortly.

ChuckL commented:
@gloomykid I think I fixed the issue.  Would you care to try it?

gloomykid commented:
@ChuckL I just tried it, works perfectly fine now! Thank you a bunch!

ChuckL commented:
Awesome!  Thanks for testing.

ChuckL commented:
Ready for review again.

Fixed
* Firefox not failing to upload or authenticate due to Firefox messing with Content-Type headers.
* XCloud not being able to restore if an extension is no longer in XKit (ie no_sidebar)
* Username starting with a number no longer mess up loading.

0xazure commented:
Finally getting around to testing this out.

Initial impressions:
- Sync and restore seem to work just fine
- Flow from the XKit side of things is the same as usual
- Trying to register with a username that already exists results in a slightly unhelpful message.  Note that the sub-message is valid and useful, '**undefined**' is not.
![screen shot 2015-09-25 at 22 03 58](https://cloud.githubusercontent.com/assets/1041456/10115368/8919bab2-63d2-11e5-855a-60f1ec4b28ad.png)

Portal thoughts:
- UI design on the XCloud portal is a bit lackluster
- Visiting the [XCloud Migration Guide](https://portal.new-xkit.com/assets/migration/xcloud-migration.html) traps the user as there is no option other than the browser's 'back' button to return to where they came from
- The ToC on the migration guide does not link to the relevant sections
- Extension size should probably also include a human-readable size in addition to the actual size in bytes ( i.e. 4.8Kb (4908 bytes) )
- Download Preferences should provide an indication that the file is not human-readable
  - Incidentally, what format _is_ `xcloud_preferences.txt`?  It doesn't appear to be valid base64.

Finally:

Slightly confused that GitHub claims this branch can be merged; I attempted to merge in `master` locally so I could get working `gulp build:extensions` instead of copy/pasting into the editor and had to resolve a conflict in `.gitignore`.

All in all, super-big :+1: @ChuckL, awesome work on this.  I'll keep playing around with it this weekend and see if I can come up with anything further to add or any ways I manage to break things.

ThePsionic commented:
Slight breaking point I noticed just now: When the data of an extension is too big, a popup will come as usual stating the extensions that cannot be backed up. If you choose "Cancel" rather than "Proceed" (or whatever it was, I forget the exact terminology), XCloud gets stuck in an endless loop and doesn't actually seem to be backing up anything, while I was expecting to just back out of the process.

ChuckL commented:
@0xazure 
The download preferences is base64 encoded with XCS prepended and XCE appended.  So if you delete those then you'll get the base64 encoded json string.  I kept the XCS and XCE for backwards compatibility.

@ThePsionic 
Ah fun,  never took the extension payload past the maximum.
One thing we might be able to do is also have local import export as well as cloud.

I'll fix up those changes that you to mentioned.  I'll have some spare cycles to work on this tonight and tomorrow.

0xazure commented on Extensions/xcloud.js:
```diff
@@ -1,7 +1,7 @@
 //* TITLE XCloud **//
-//* VERSION 0.2 REV B **//
+//* VERSION 0.3.0 **//
```
Can we make the version `1.0.0` now?  No reason to keep it pre-1.0, especially given this is a rather major change.

0xazure commented on Extensions/xcloud.js:
```diff
@@ -1,7 +1,7 @@
 //* DESCRIPTION Sync XKit data on clouds **//
-//* DETAILS XCloud stores your XKit configuration on STUDIOXENIX servers so you can back up your data and synchronize it with other computers and browsers easily. **//
+//* DETAILS XCloud stores your XKit configuration on New-XKit servers so you can back up your data and synchronize it with other computers and browsers easily. Also compatable with STUDIOXENIX servers.**//
 //* DEVELOPER STUDIOXENIX **//
```
Should probably update the developer field to `new-xkit`.

0xazure commented on Extensions/xcloud.js:
```diff
@@ -551,77 +684,56 @@ XKit.extensions.xcloud = new Object({
 
 	start_upload: function() {
 
 		// XKit.extensions.xkit_preferences.close();
```
Can we remove this since we're not using it?

0xazure commented:
Now that the SSL/TLS certificates on the old XKit servers are dead, I don't know how viable the 'use old servers' option is anymore.

Other than that, this looks good to merge, to me.  We'll have to coordinate the deployment and announcement posts before we rebuild and make this go live.

Wolvan commented:
@0xazure the old servers still work as they did before the depcretation as long as you visit https://www.xkitcs.com/ once and add it as a security exception

ChuckL commented:
I think we should just leave that functionality in. Like Wolvan said, there are work arounds and if the certs are valid again I don't see a reason to restrict that functionality.   I'm ready to merge whenever, my server is ready.

0xazure commented:
I don't find requiring end-users to add a security exception for an expired cert to be part of an acceptable user flow.  The currently implementation was fine before the certs expired because it required no user intervention to get it to work.  'Functionality' that requires the _end-user_ to perform a work-around should not be presented in this manner.

I understand the desire to maintain this functionality and it would be nice for those users that want to jump through hoops to get it to work to make it available, but I disagree with how it's being presented to the end-user in the form of just a checkbox with no links or tooltips about the steps necessary to get it to work.

Remember that if XCloud can't connect to a server, XKit gets stuck in that state and the user is required to refresh the page to reset the state.  Also note that per [#574 (comment)](https://github.com/new-xkit/XKit/pull/574#issuecomment-147554588) any work-around regarding adding security exceptions may not be functional on Safari.

ChuckL commented:
Since the main concern was userflow, I set up an AWS API Gateway which acts as a proxy to the old XCloud servers which fixes all of the security exceptions.  In doing so I also found a way to register on the old XCloud servers as well (D'oh)

Anyways, the XKit getting stuck in that state is clearly a bug which I'll probably work on fixing.  However that bug seems minor and I would suggest merging in XCloud without that bug fixed since I think the cycles I can dedicate to XKit are dwindling :<

Anyways, I hope this is an acceptable solution.

0xazure commented:
Tested things out, and it works beautifully.  No issues on the user's side; excellent call on the AWS API Gateway.  I was never against the functionality to begin with, rather that it should be as simple as possible for the end-user to use it.  Accepting out of date certificates isn't all that user friendly.

@ChuckL thank you so much for your work on this.  Massive effort on your part, it's greatly appreciated.

As far as XCloud getting stuck, it's been like that in XKit for quite a while, so it's not imperative that it gets fixed in this cycle.

@homu r+

homu commented:
:pushpin: Commit 9aada10 has been approved by `0xazure`

<!-- @homu r=0xazure 9aada10dda3013ce6fa75895406caa07f7ace9ed -->

homu commented:
:hourglass: Testing commit 9aada10 with merge 9d86490...

homu commented:
:sunny: Test successful - [status](https://travis-ci.org/new-xkit/XKit/builds/86319678)
